### ISSUE TITLE
Should we allow the same app to bind to the same service more than once?

### ISSUE THREAD

**Issue by duglin** [kubernetes-incubator/service-catalog#793]
I actually don't think we can stop this given how we use label selectors to define "the app".
But we may want to modify our docs to call this out since this is different than what CF does.

**Comment by krancour** [comment]
I'm about to get philosophical...

imo, "app" has an increasingly loose definition-- especially in the world of micro-services. Is a single micro-service an app? Some may say yes. Is a set of micro-services that form a cohesive system an app? Some may say yes. Are they _both_ apps? Sure. (Or no.)

(I'll qualify the above by saying that in some platforms, the meaning of "app" is not so open to interpretation, but vis a vis plain k8s, I think "app" is pretty nebulous. After all-- there is no `App` type, so as far as k8s is concerned, "app" isn't even a thing. It's a human construct.)

Anyway. My point is that if we can't definitively say what _is_ or _isn't_ an app, we can't (or shouldn't?) put restrictions on this. In my mind, it's perfectly legit that someone might consider a namespace to be the boundaries of their app and that n micro-services within that namespace might all require interaction with one common service instance. Couldn't they all share one binding? They could. That's what I would do... until I had a reason not to. Such reasons probably do exist.

Bottom line is that, imho, I wouldn't presume it to be our place to _say_ whether one "app" can have multiple bindings to one service instance or not.

**Comment by duglin** [comment]
Overall, as I said in my opening comment, I tend to agree for K8s.  However, let's say we get to the point where we support something like:
```
kubectl create-instance mongo mydb
kubectl service-bind mydb deployment/myapp
```
(ping @simonleung8 since you're working on our cli ux)

meaning, create a mydb instance of mongo and bind it to the deployment called myapp - it uses the label selector from deployment/myapp as the label selector in the pod preset. Not sure if this is where we'll end up, but let's assume we manage to come up with some really nice UX like this where we don't always require people to get into the yaml every time.

So, the question is... what if they try to run the `kubectl service-bind...` command again? Do we create a 2nd binding, 2nd secret and 2nd podpreset or do we try to detect this and stop it like CF does?  Does it even make sense to allow it - even if we could stop it?  These are the questions I'd like us to ponder.  

I see little value in allowing the same service instance to be bound to the same "thing" - if we can clearly define what the "thing" is.  Seems like some times we may be able to, other we may not.

**Comment by simonleung8** [comment]
IMO, I would like creating a binding to the same 'thing' to be idempotent, which means you can call `service-bind` multiple times, but the same 1st binding is returned. I don't currently see a reason to prevent 2nd binding other than to avoid creating duplicated 2nd bindings.
For the first iteration of the UX, I propose we go with what is simple, which I suspect it to be creating multiple bindings when multiple `service-bind` is called. We can change the behavior later depending on our needs.

**Comment by krancour** [comment]
I'm almost afraid to ask... there's lots in this thread that is scaring me...

First, @duglin:

> ```
> kubectl create-instance mongo mydb
> kubectl service-bind mydb deployment/myapp
> ```

`create-instance` and `service-bind` aren't legitimate `kubectl` sub-commands.

Possibilities:

1. You said "kubectl"  in error. You intended to name some other (tbd) CLI binary name?
1. We think we're going to _expand_ `kubectl`. (I really hope not.) `kubectl`, fundamentally, is just an interface for CRUD on resources. That should make it compatible with _all_ aggregated APIs. Adding imperative sub-commands that are specific to any given addon, such as ours, seems ill-advised. (Unless there is some special mechanism for extending the `kubectl` CLI with plugins or something. If there is, I am unaware of it.)

Second thing that's concerning me:

> I would like creating a binding to the same 'thing' to be idempotent

That's not at all for us to decide. This is a broker implementation detail.

**Comment by krancour** [comment]
@arschles has told me kubectl plugins are coming in 1.7. So, I assume that at least partially addresses some of my concerns. I'd still like to discuss further, however.

**Comment by duglin** [comment]
re: kubectl - it was a little bit of a placeholder name for whatever we decide to expose w.r.t. some nicer porcelain exe, and a little bit of an acknowledgement to the fact that kubectl plugins exist, just in a different syntax. The bigger point was that we may have a UX that doesn't necessary expose yaml (label selectors) directly, which means we may have an opportunity in some cases to actually have a "thing" (app) that is more than an abstract entity.

re: "same thing" - the broker only gets a little bit of control of this. If we send the same createBidning() request (meaning the same bindingID) then if the broker returns anything at all I would expect the same creds as before since its the same binding resource. But, its up to the platform (kube) to decide when to change create a new bindingID. So, in the example I gave above, if someone does `kubectl service-bind mydb deployment/myapp` twice in a row then Kube/we get to decide whether that generates a new bindingID or not.

Kind of related to this... is the entire notion of the appID and "context" object that's being proposed for the spec.  Its very possible that even if the bindingID changes on each createBinding() request the broker could still look at things like the org/space/appID and decide to change (or not) the creds - or even reject it. For example, the broker could look at the org/space/appID and decide its the same thing as before and act accordingly.  Now, if we continue on our current path where we send the namespace name as the appID then that makes the granularity of those decisions very coarse and not nearly as fine-grained as they are in CF.  This is something else we may need to think about in this discussion.  For example, one option would be to make the appID be a hash of the namespace+labelSelector of the podPreset.  Just brainstorming though, needs a lot more thinking....

**Comment by pmorie** [comment]
I think this is entirely related to what we consider an 'app', ie, this depends on #798

**Comment by pmorie** [comment]
I think we have precedent for the use-case of:

Given that I know what an 'app' is,

As a developer using the catalog, I want to bind to the same app multiple times with different parameters so that I can have multiple connections with different properties

Example: the service is a database, and you want a read-only connection and a read-write connection

**Comment by duglin** [comment]
Was going thru old FF tabs and came across this... one thing that could influence this is that it seems to me we could technically trace things back from podpresets to the secret to the binding to the instance. Which means, for any one pod it might be possible to detect the same instance being bound twice - if we wanted to.  Its also possible that, if defaults are chosen, that the podpreset logic would naturally detect this as an error because there might be some conflict between the two podpresets.    Maybe that's the answer... let podpreset detect it in those obvious (default) cases, but people could get around it if they really wanted by giving explicit names/values in the right spots.

**Comment by pmorie** [comment]
@duglin circling back - do you think this is still an issue?

**Comment by duglin** [comment]
Not 100% sure but my gut is telling me that if nothing else we'll need some kind of doc update to talk about this. I'd like to wait until we figure out 1) what we send for the "app_id", and 2) what we're going to do with PodPreset/Bindings in this space before we make a final decision on this one.

**Comment by pmorie** [comment]
I wonder if the most conservative thing to do at this point is to stop sending it for now, document that we don't send it, and begin sending it when we know more about what it means.

Thoughts?

**Comment by duglin** [comment]
Blocked by https://github.com/kubernetes-incubator/service-catalog/issues/818

**Comment by duglin** [comment]
given #818 was CWNA, for now, we've decided that Bindings have no notion of apps, so therefore thinking about trying to block the same app from binding to the same service would be out of scope right now.

Proposal: CWNA - can reopen if things change or we get new data.  Will add to the agenda for monday's design session.

**Comment by MHBauer** [comment]
What's a CWNA?

**Comment by duglin** [comment]
Close With No Action

**Comment by duglin** [comment]
on today's call we agreed to cwna.
